### PR TITLE
Fix max backward counts for low-precision dtypes

### LINKED ISSUE #12296: max backward can underflow
```
from tinygrad import Tensor
t = Tensor.ones(70000, dtype="half", requires_grad=True).contiguous()
t.max().backward()
print(t.grad.tolist()[:10])
```
gives `[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]`

https://github.com/tinygrad/tinygrad/blob/476a2a0a962fa8a145c612f63b72eeb4dc3a84f3/tinygrad/gradient.py#L10 uses the gradient dtype for counting

### PR DESCRIPTION
**Pull request by markerdmann**
- Fixes #12296 by keeping mask counts for Tensor.max().backward() in safe dtypes (int32/float32) so large groups of equal maxima no longer underflow in half/bfloat16. The mask stays boolean, converts to int32 for counting, divides in float32, then casts back to the incoming gradient dtype (tinygrad/gradient.py:7-
    14).

- Adds TestTinygrad.test_max_backward_half_many_equal_maxima to reproduce the former failure and assert finite, positive gradients that sum to ~1 even for 70k half-precision maxima (test/test_tensor.py:83-89).

- ./.venv/bin/python -m pytest -q test/test_tensor.py and the targeted -k max_backward_half_many_equal_maxima pass after the change.

### CODE REVIEW

**Comment by markerdmann** [comment]
@chenyuxyz How does this look so far? 

This is a fix for https://github.com/tinygrad/tinygrad/issues/12296

**Comment by chenyuxyz** [comment]
importing dtype in `gradient.py` looks weird. can it be more fundamental?

**Comment by markerdmann** [comment]
> importing dtype in `gradient.py` looks weird. can it be more fundamental?

@chenyuxyz Thanks for the feedback. I reworked the MAX backward rule to use the standard accumulation dtype (sum_acc_dtype/least_upper_float) so the count/normalize happens in the shared ‘safe acc’ dtype before casting back to the grad dtype. This keeps the policy centralized and still fixes the half underflow case. Is this in line with what you were thinking, or did you want it to be even more fundamental?

**Comment by markerdmann** [comment]
@chenyuxyz Just checking in on this one! Let me know if you'd like further improvements.
